Spanner/src/Session/CacheSessionPool.php

```diff
@@ -851,4 +852,123 @@ private function manageSessionsToDelete(array &$data)
+            for ($expiredPos = 0; $expiredPos < $len; $expiredPos++) {
+                if ($queue[$expiredPos]['expires'] > $now) {
+                    break;
+                }
+            }
+            for ($oldPos = $expiredPos; $oldPos < $len; $oldPos++) {
+                if ($queue[$oldPos]['expires'] > $oldThreshold) {
+                    break;
+                }
+            }
+            $freshPos = $len - 1;
+            if (isset($prevMaintainTime)) {
+                $freshThreshold = $prevMaintainTime + self::SESSION_EXPIRATION_SECONDS;
+                for (; $freshPos >= 0; $freshPos--) {
+                    if ($queue[$freshPos]['expires'] <= $freshThreshold) {
+                        break;
+                    }
+                }
+            }
+            $freshCount = $len - 1 - $freshPos;
+            $oldQueue = array_splice($queue, $expiredPos, ($oldPos - $expiredPos));
+            array_splice($queue, 0, $expiredPos);
+            $extraQueue = [];
+
+            $totalCount = count($data['inUse']) + count($queue) + count($oldQueue);
+            $minCount = $this->config['minSessions'];
+            $extraCount = ($totalCount - $minCount);
+            if ($extraCount > 0) {
+                $extraQueue = array_splice($oldQueue, -$extraCount);
+            }
+
+            foreach ($oldQueue as $item) {
+                $session = $this->database->session($item['name']);
+                if ($this->refreshSession($session)) {
+                    $queue[] = [
+                        'name' => $item['name'],
+                        'expires' => $session->expiration(),
+                    ];
+                    $freshCount++;
+                } else {
+                    $totalCount--;
+                }
+            }
+
+            if (isset($prevMaintainTime)) {
+                $maintainInterval = $now - $prevMaintainTime;
+                $maxLifetime = self::SESSION_EXPIRATION_SECONDS - 600;
+                $meanRefreshCount = (int)(min($totalCount, $minCount) * $maintainInterval / $maxLifetime);
+                if ($meanRefreshCount > $minCount) {
+                    $meanRefreshCount = $minCount;
+                }
+                $refreshCount = $meanRefreshCount - $freshCount;
+                if ($refreshCount > 0) {
+                    $count = min($refreshCount, count($queue));
+                    for ($pos = 0; $pos < $count; $pos++) {
+                        $item = $queue[$pos];
+                        $session = $this->database->session($item['name']);
+                        if ($this->refreshSession($session)) {
+                            $queue[] = [
+                                'name' => $item['name'],
+                                'expires' => $session->expiration(),
+                            ];
+                        }
+                    }
+                    array_splice($queue, 0, $count);
+                }
+            }
```

Comment:
This is also quite subtle and it needs some documentation. Especially the fact that you're refreshing some sessions from the front of the queue and then moving it back to the queue. So next time around, it will be a new set of sessions that will be refreshed.